File: Python/beginner/01_Sorting/QuickSort.py
# 퀵정렬

arr = list(map(int, input().split()))

def quick_sort(arr, start, end):
    if start >= end:
        return

    pivot = start
    left = start + 1
    right = end 

    while left <= right:
        while arr[pivot] >= arr[left] and left <= end:
            left += 1

        while arr[pivot] <= arr[right] and right > start:
            right -= 1

        if left > right:
            arr[right], arr[pivot] = arr[pivot], arr[right]
        else:
            arr[left], arr[right] = arr[right], arr[left]

    quick_sort(arr, 0, right - 1)
    quick_sort(arr, right + 1, end)


def pythonic_quick_sort(arr):
    if len(arr) <= 1:
        return arr
    
    pivot = arr[0]
    tail = arr[1:]

    left = [x for x in tail if x <= pivot]
    right = [x for x in tail if x > pivot]

    return pythonic_quick_sort(left) + [pivot] + pythonic_quick_sort(right)

sorted_arr = pythonic_quick_sort(arr)
print(sorted_arr)

File: Kotlin/src/main/kotlin/beginner/a_Sorting/quick_sort.kt
package beginner.a_Sorting

// quick sort with kotlin style
// 매번 새로운 리스트를 만드므로 비효율적 ㅠㅠ
fun solve(arr: IntArray): IntArray {
    if (arr.size < 2)
        return arr

    val pivot = arr[arr.size / 2]
    val left = arr.filter { it < pivot }.toIntArray()
    val right = arr.filter { it > pivot }.toIntArray()

    return solve(left) + intArrayOf(pivot) + solve(right)
}

fun quick_sort() = with(System.`in`.bufferedReader()) {
    val arr = readLine()
        .split(" ")
        .map { it.toInt() }
        .toIntArray()

    val sorted_arr = solve(arr)

    println("[ 정렬 완료 ]")
    for (k in sorted_arr)
        print("$k ")
}